data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java

```diff
@@ -217,13 +218,37 @@ private <T> Record<Event> getRecord(ConsumerRecord<String, T> consumerRecord) {
                 value = objectMapper.readValue(jsonParser, Map.class);
             } else if (schema == MessageFormat.PLAINTEXT) {
                 value = (String)consumerRecord.value();
+                plainTextMode = true;
+            } else if (schema == MessageFormat.JSON) {
+                value = objectMapper.convertValue(value, Map.class);
             }
         } catch (Exception e){
             LOG.error("Failed to parse JSON or AVRO record", e);
         }
-        data.put(key, value);
-
+        if (!plainTextMode) {
+            if (!(value instanceof Map)) {
+                data.put(key, value);
+            } else {
+                Map<String, Object> valueMap = (Map<String, Object>)value;
+                if (kafkaKeyMode == KafkaKeyMode.INCLUDE_AS_FIELD) {
+                    valueMap.put("kafka_key", key);
```

Comment:
minor: topic_key could be better name. Would customer want to have this configurable also ?

Reply:
Key could be different for each message. So, it cannot be "topic_key". It's the message key from kafka.